src/audio/asrc/asrc.c

```diff
@@ -836,11 +831,17 @@ static int asrc_copy(struct comp_dev *dev)
 	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
 			       source_list);
 
+	buffer_lock(source);
+	buffer_lock(sink);
```

Comment:
should we have an API here to guarantee ordering (and to avoid deadlocks).
i.e.
```c
buffer_lock_source_sink(source, sink);
...
buffer_unlock_source_sink(source, sink);
```
This may need to be atomic too.

Reply:
Ordering doesn't really matter. We just need to have it locked before getting buffer's current parameters like w_/r_ptr, free, avail etc. It just happens that both of these buffers are locked in the same place. I think adding additional functions is a bit misleading in this particular case. However we need to think in the near future about some kind of additional wrappers for getting/releasing buffers to avoid explicit lock/unlock calls.

Reply:
What happens if
```
lock(source);
IRQ -> lock(sink);->lock(source) -> DEADLOCK
```
or
```
lock(source);
lock(sink);
update(source);
IRQ -> update(source); ->done <<<< just trashed source
update(sink);
unlock(source);
unlock(sink);
```
We should really make this atomic.

Reply:
@lgirdwood You are rather suggesting that we should also block irq, because we can get preempted, but still there is no correlation between locking source and sink buffers, since they are totally separate instances. As for preemption, we are setting highest irq lvl on the xtensa arch level when executing any interrupt, so technically it won't happen, but we can add this to make it more generic for other architectures in the future.